lib/galaxy/managers/tools.py

```diff
@@ -30,31 +47,51 @@
     from galaxy.managers.base import OrmFilterParsersType
 
 
+def tool_payload_to_tool(app, tool_dict: Dict[str, Any]) -> Optional[Tool]:
+    tool_source = YamlToolSource(tool_dict)
+    tool = create_tool_from_source(app, tool_source=tool_source, tool_dir=None)
+    return tool
+
+
 class DynamicToolManager(ModelManager):
     """Manages dynamic tools stored in Galaxy's database."""
 
     model_class = model.DynamicTool
 
+    def get_tool_by_id_or_uuid(self, id_or_uuid: Union[int, str]):
+        if isinstance(id_or_uuid, int):
+            return self.get_tool_by_id(id_or_uuid)
+        else:
+            return self.get_tool_by_uuid(id_or_uuid)
+
     def get_tool_by_uuid(self, uuid: Optional[Union[UUID, str]]):
         stmt = select(DynamicTool).where(DynamicTool.uuid == uuid)
         return self.session().scalars(stmt).one_or_none()
 
     def get_tool_by_tool_id(self, tool_id):
-        stmt = select(DynamicTool).where(DynamicTool.tool_id == tool_id)
+        stmt = select(DynamicTool).where(DynamicTool.tool_id == tool_id, DynamicTool.public == true())
+        return self.session().scalars(stmt).one_or_none()
+
+    def get_unprivileged_tool_by_uuid(self, user: model.User, uuid: Union[UUID, str]):
+        stmt = self.owned_unprivileged_statement(user).where(DynamicTool.uuid == uuid)
+        return self.session().scalars(stmt).one_or_none()
+
+    def get_unprivileged_tool_by_tool_id(self, user: model.User, tool_id: str):
+        stmt = self.owned_unprivileged_statement(user).where(DynamicTool.tool_id == tool_id)
         return self.session().scalars(stmt).one_or_none()
 
     def get_tool_by_id(self, object_id):
-        stmt = select(DynamicTool).where(DynamicTool.id == object_id)
+        stmt = select(DynamicTool).where(DynamicTool.id == object_id, DynamicTool.public == true())
         return self.session().scalars(stmt).one_or_none()
 
-    def create_tool(self, trans, tool_payload, allow_load=True):
+    def create_tool(self, trans: ProvidesUserContext, tool_payload: DynamicToolPayload, allow_load=True):
```

Comment:
```suggestion
    def create_tool(self, tool_payload: DynamicToolPayload, allow_load=True):
```

Reply:
We'll eventually have to do access permission checks, those would consume trans ... I will refactor this once that is there.

Reply:
Just ignore the suggestions then, sorry for the noise!